### PR TITLE
Add set() overrides to lite sdk

### PR DESCRIPTION
**Pull request by thebrianchen**
Setting the base as #3254 since it requires the changes to `applyFirestoreDataConverter`. Since this branch sits on top of #3254, I will diff this against master once #3254 goes in.

### CODE REVIEW

File: packages/firestore/lite/index.d.ts [hunk 1]
```diff
@@ -45,6 +45,7 @@ export function setLogLevel(logLevel: LogLevel): void;
 
 export interface FirestoreDataConverter<T> {
   toFirestore(modelObject: T): DocumentData;
+  toFirestore(modelObject: Partial<T>, options: SetOptions): DocumentData;
   fromFirestore(snapshot: QueryDocumentSnapshot): T;
```
Comment (schmidt-sebastian) [inline packages/firestore/lite/index.d.ts:49]: Should we explicitly type this as QueryDocumentSnapshot<DocumentData> to make it easier to understand?

File: packages/firestore/lite/test/integration.test.ts [hunk 2]
```diff
@@ -1045,4 +1076,111 @@ describe('withConverter() support', () => {
...
+        batch.set(ref, { title: 'olive' }, { merge: true })
+      ).to.throw(
+        'Function WriteBatch.set() called with invalid data ' +
+          '(via `toFirestore()`). Unsupported field value: undefined ' +
+          '(found in field author in document posts/post)'
+      );
+    });
+  });
+
+  it('WriteBatch.set() supports partials with merge', async () => {
```
Comment (schmidt-sebastian) [inline packages/firestore/lite/test/integration.test.ts:1095]: Can you move these tests to https://github.com/firebase/firebase-js-sdk/blob/master/packages/firestore/lite/test/integration.test.ts#L253, which allows you to write one test for WriteBatch, Transaction and set().

File: packages/firestore/lite/index.d.ts [hunk 3]
```diff
@@ -45,6 +45,7 @@ export function setLogLevel(logLevel: LogLevel): void;
 
 export interface FirestoreDataConverter<T> {
   toFirestore(modelObject: T): DocumentData;
+  toFirestore(modelObject: Partial<T>, options: SetOptions): DocumentData;
```
Comment (schmidt-sebastian) [inline packages/firestore/lite/index.d.ts:48]: Note that this change also need to be made in exp/index.d.ts

File: packages/firestore/lite/index.d.ts (hunk 1 above)
Comment (thebrianchen) [reply packages/firestore/lite/index.d.ts]: done.

File: packages/firestore/lite/index.d.ts (hunk 3 above)
Comment (thebrianchen) [reply packages/firestore/lite/index.d.ts:48]: done for converter and `SetOptions`.

File: packages/firestore/lite/test/integration.test.ts (hunk 2 above)
Comment (thebrianchen) [reply packages/firestore/lite/test/integration.test.ts]: done.

**Pull request merged**